src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsAllShardsMetricsCollector.java

```diff
@@ -0,0 +1,251 @@
+            NodeStatsAllShardsMetricsCollector.class).samplingInterval;
+    private static final int KEYS_PATH_LENGTH = 2;
+    private static final Logger LOG = LogManager.getLogger(NodeStatsAllShardsMetricsCollector.class);
+    private HashMap<String, IndexShard> currentShards;
+    private final PerformanceAnalyzerController controller;
+
+
+    public NodeStatsAllShardsMetricsCollector(final PerformanceAnalyzerController controller) {
+        super(SAMPLING_TIME_INTERVAL, "NodeStatsMetrics");
+        currentShards = new HashMap<>();
+        this.controller = controller;
+    }
+
+    private void populateCurrentShards() {
+        currentShards.clear();
+        currentShards = Utils.getShards();
+    }
+
+    private Map<String, ValueCalculator> valueCalculators = new HashMap<String, ValueCalculator>() { {
+        put(ShardStatsValue.INDEXING_THROTTLE_TIME.toString(),
+                (shardStats) -> shardStats.getStats().getIndexing().getTotal().getThrottleTime().millis());
+
+        put(ShardStatsValue.CACHE_QUERY_HIT.toString(), (shardStats) -> shardStats.getStats().getQueryCache().getHitCount());
+        put(ShardStatsValue.CACHE_QUERY_MISS.toString(), (shardStats) -> shardStats.getStats().getQueryCache().getMissCount());
+        put(ShardStatsValue.CACHE_QUERY_SIZE.toString(), (shardStats) -> shardStats.getStats().getQueryCache().getMemorySizeInBytes());
+
+        put(ShardStatsValue.CACHE_FIELDDATA_EVICTION.toString(), (shardStats) -> shardStats.getStats().getFieldData().getEvictions());
+        put(ShardStatsValue.CACHE_FIELDDATA_SIZE.toString(), (shardStats) -> shardStats.getStats().getFieldData().getMemorySizeInBytes());
+
+        put(ShardStatsValue.CACHE_REQUEST_HIT.toString(), (shardStats) -> shardStats.getStats().getRequestCache().getHitCount());
+        put(ShardStatsValue.CACHE_REQUEST_MISS.toString(), (shardStats) -> shardStats.getStats().getRequestCache().getMissCount());
+        put(ShardStatsValue.CACHE_REQUEST_EVICTION.toString(), (shardStats) -> shardStats.getStats().getRequestCache().getEvictions());
+        put(ShardStatsValue.CACHE_REQUEST_SIZE.toString(), (shardStats) -> shardStats.getStats().getRequestCache().getMemorySizeInBytes());
+
+    } };
+
+    @Override
+    public String getMetricsPath(long startTime, String... keysPath) {
+        // throw exception if keysPath.length is not equal to 2 (Keys should be Index Name, and ShardId)
+        if (keysPath.length != KEYS_PATH_LENGTH) {
+            throw new RuntimeException("keys length should be " + KEYS_PATH_LENGTH);
+        }
+        return PerformanceAnalyzerMetrics.generatePath(startTime, PerformanceAnalyzerMetrics.sIndicesPath, keysPath[0], keysPath[1]);
+    }
+
+    @Override
+    public void collectMetrics(long startTime) {
+        IndicesService indicesService = ESResources.INSTANCE.getIndicesService();
+
+        if (indicesService == null) {
+            return;
+        }
+
+        try {
+            populateCurrentShards();
```

Comment:
we are populating this shard haspmap twice per collection (1 per each collector). Is there any specific reason to run two collectors in separate threads? Can we combine those two collector to reduce memory footprint. We should be conservative here as PA itself needs to be lightweighted

Reply:
This is just populating shard IDs from the inbuilt shard Iterator from ES. But I agree we should save memory wherever possible. This can be taken up as a separate task.